src/test/optimizer/strategy/predicate_reordering_rule_test.cpp

```diff
@@ -240,7 +240,9 @@ TEST_F(PredicateReorderingTest, PredicatesWithMultipleOutputs) {
   EXPECT_EQ(predicate_b_node->left_input(), table_node);
 }
 
-TEST_F(PredicateReorderingTest, SimpleValidateReorderingTest) {
+// TODO
+// Currently disabled, when MatchesAll gives better performance we should consider removing the rule
+/* TEST_F(PredicateReorderingTest, SimpleValidateReorderingTest) {
```

Comment:
To resolve this, I'd say we compare release performance and then decide which ordering is currently better?

However, strangely, master doesn't compile for me on the VM with release so our branch doesn't either:
```
In file included from /home/Richard.Ebeling/hyrise/src/lib/storage/vector_compression/fixed_size_byte_aligned/fixed_size_byte_aligned_vector.hpp:9,
                 from /home/Richard.Ebeling/hyrise/src/lib/storage/vector_compression/resolve_compressed_vector_type.hpp:7,
                 from /home/Richard.Ebeling/hyrise/src/lib/storage/dictionary_segment/dictionary_segment_iterable.hpp:9,
                 from /home/Richard.Ebeling/hyrise/src/lib/storage/create_iterable_from_segment.ipp:3,
                 from /home/Richard.Ebeling/hyrise/src/lib/storage/create_iterable_from_segment.hpp:82,
                 from /home/Richard.Ebeling/hyrise/src/lib/operators/join_nested_loop.cpp:9:
/home/Richard.Ebeling/hyrise/src/lib/storage/vector_compression/fixed_size_byte_aligned/fixed_size_byte_aligned_decompressor.hpp: In function ‘void {anonymous}::join_two_typed_segments(const BinaryFunctor&, LeftIterator, LeftIterator, RightIterator, RightIterator, opossum::ChunkID, opossum::ChunkID, const opossum::JoinNestedLoop::JoinParams&) [with BinaryFunctor = std::greater_equal<void>; LeftIterator = opossum::AnySegmentIterator<int>; RightIterator = opossum::FrameOfReferenceSegmentIterable<int>::PointAccessIterator<opossum::FixedSizeByteAlignedDecompressor<unsigned char> >]’:
/home/Richard.Ebeling/hyrise/src/lib/storage/vector_compression/fixed_size_byte_aligned/fixed_size_byte_aligned_decompressor.hpp:19:46: error: ‘*((void*)(& right_it)+32).opossum::FixedSizeByteAlignedDecompressor<unsigned char>::_data’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
   19 |   uint32_t get(size_t i) final { return _data[i]; }
      |                                         ~~~~~^
/home/Richard.Ebeling/hyrise/src/lib/operators/join_nested_loop.cpp:51:15: note: ‘*((void*)(& right_it)+32).opossum::FixedSizeByteAlignedDecompressor<unsigned char>::_data’ was declared here
   51 |     for (auto right_it = right_begin; right_it != right_end; ++right_it) {
      |               ^~~~~~~~
```
(GCC, release build, same error repeats 17 times)

Reply:
Strange, master should compile without problems. You did not change any of these files, did you?

Reply:
41d70ad904 works for me with GCC (g++ (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008) in debug and release, `hyriseTest`. `CXXFLAGS=-fdiagnostics-color=always cmake -DENABLE_NUMA_SUPPORT=Off -DCMAKE_BUILD_TYPE=Release -G Ninja ..`.

Reply:
I think I didn't -- but maybe I'm missing something. If you have some time, feel free to
```
sudo su -- Richard.Ebeling
cd ~/hyrise/build-release
make
```
to reproduce on vm-nordes.

Reply:
Might not be you: https://stackoverflow.com/questions/14132898/gcc-wuninitialized-wmaybe-uninitialized-issues

Do you have the same compiler version?

Reply:
I have the exact same compiler version as @Bensk1. If also fails in a completely new build environment:
```bash
mkdir ~/test && cd ~/test
git clone git@github.com:Xperianer/hyrise.git
cd hyrise
mkdir build && cd build
git submodule update --init
git checkout polymorphic-poslist-compiling
git log  # is on 41d70ad
cmake .. -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DCMAKE_BUILD_TYPE=Release -DENABLE_NUMA_SUPPORT=Off
make -j 32 hyriseTest
```

same thing happens when building with ninja and the command @Bensk1 gave:
```bash
CXXFLAGS=-fdiagnostics-color=always cmake -DENABLE_NUMA_SUPPORT=Off -DCMAKE_BUILD_TYPE=Release -G Ninja ..
ninja hyriseTest
```

`/usr/bin/env` does not report any environment variables that should change gcc's behaviour.
Is there any other place where I could have globally configured gcc to behave differently?

Reply:
I'd like to test something in your environment. Let me know when it's free.

Reply:
Go ahead :)

Reply:
My guess is that it is another gcc unused bug. Reasons for that:
* clang isn't unhappy about it
* `_data` is a reference and, as such, it should be impossible to create an object where it is uninitialized. You should get "error: constructor for 'X' must explicitly initialize the reference member 'foo'" even without any warning messages enabled
* GCC's unused detection is known to be buggy

My suggestion is to add this:

```
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
...
#pragma GCC diagnostic pop
```

Hopefully, doing this in the get() method should be sufficient.

Reply:
What you suggested works for me. Benchmarks on whether to do validate reordering are still todo.

Reply:
I guess you have an outdated Cmake version. I've seen the problem several times with an old CMake version and a new version fixed it. But I never took a deeper look into that.

Reply:
That sounds super weird.

Reply:
It does ... shouldn't be related. But it solved the issue on two machines for us.

Reply:
Cmake was already the most recent version in the ubuntu repo (3.13.4). The error still occurs if we remove the pragmas.

Reply:
Yes, Ubuntu lacks behind here. See our docker file for an example. In the docker setup, we manually pull the latest cmake version.
@Bensk1, @mrks : any reason not to update vm-nordes and see if that changes something for good?

Reply:
Unless we add that manually build cmake version as a hard requirement (which I don't see a reason too and which will also mess with those IDE things that people seem to use) that won't change anything. I'd stick with the pragmas.

Reply:
But could we than at least open an issue and add a todo to the pragma to remove it as soon as CMake is updated on Ubuntu?

Reply:
Last time I made Ubuntu's cmake version mandatory, some people complained that this would not work with their IDEs.

Reply:
So what do we do here? I'd be fine with
> But could we than at least open an issue and add a todo to the pragma to remove it as soon as CMake is updated on Ubuntu?

Reply:
As mentioned above, this is not solely an issue with the CMake version shipped with Ubuntu, but also that within IDEs. I'd simply leave the pragma it and be done with it. It's not that it really hurts anyone if it remains in there for a while.